src/peer-store/persistent/index.js

```diff
@@ -0,0 +1,225 @@
+    await this._commitData()
+    log('Persistent PeerStore stopped')
+  }
+
+  /**
+   * Add modified peer to the dirty set
+   * @private
+   * @param {Object} params
+   * @param {PeerId} params.peerId
+   */
+  _addDirtyPeer ({ peerId }) {
+    const peerIdstr = peerId.toB58String()
+
+    log('add dirty peer', peerIdstr)
+    this._dirtyPeers.add(peerIdstr)
+
+    if (this._dirtyPeers.size >= this.threshold) {
+      // Commit current data
+      this._commitData()
+    }
+  }
+
+  /**
+   * Add all the peers current data to a datastore batch and commit it.
+   * @private
+   * @param {Array<string>} peers
+   * @return {Promise<void>}
+   */
+  async _commitData () {
+    const commitPeers = Array.from(this._dirtyPeers)
+
+    if (!commitPeers.length) {
+      return
+    }
+
+    // Clear Dirty Peers set
+    this._dirtyPeers.clear()
+
+    log('create batch commit')
+    const batch = this._datastore.batch()
+    for (const peerIdStr of commitPeers) {
+      // PeerId (replace by keyBook)
+      const peerId = this.peerIds.get(peerIdStr)
+
+      // Address Book
+      this._batchAddressBook(peerId, batch)
+
+      // Proto Book
+      this._batchProtoBook(peerId, batch)
+    }
+
+    await batch.commit()
+    log('batch committed')
+  }
+
+  /**
+   * Add address book data of the peer to the batch.
+   * @private
+   * @param {PeerId} peerId
+   * @param {Object} batch
+   */
+  _batchAddressBook (peerId, batch) {
+    const b32key = peerId.toString()
+    const key = new Key(`${ADDRESS_NAMESPACE}${b32key}`)
+
+    const addresses = this.addressBook.get(peerId)
+
+    try {
+      // Deleted from the book
+      if (!addresses) {
+        batch.delete(key)
+        return
+      }
+
+      const encodedData = Addresses.encode({
+        addrs: addresses.map((address) => ({
+          multiaddr: address.multiaddr.buffer
+        }))
+      })
+
+      batch.put(key, encodedData)
+    } catch (err) {
+      log.error(err)
+    }
+  }
+
+  /**
+   * Add proto book data of the peer to the batch.
+   * @private
+   * @param {PeerId} peerId
+   * @param {Object} batch
+   */
+  _batchProtoBook (peerId, batch) {
+    const b32key = peerId.toString()
+    const key = new Key(`${PROTOCOL_NAMESPACE}${b32key}`)
+
+    const protocols = this.protoBook.get(peerId)
+
+    try {
+      // Deleted from the book
+      if (!protocols) {
+        batch.delete(key)
+        return
+      }
+
+      const encodedData = Protocols.encode({ protocols })
+
+      batch.put(key, encodedData)
+    } catch (err) {
+      log.error(err)
+    }
+  }
+
+  /**
+   * Process datastore entry and add its data to the correct book.
+   * @private
+   * @param {Object} params
+   * @param {string} params.key datastore key
+   * @param {Buffer} params.value datastore value stored
+   */
+  _processDatastoreEntry ({ key, value }) {
+    try {
+      const keyParts = key.toString().split('/')
+      const peerId = PeerId.createFromCID(keyParts[3])
```

Comment:
I don't think this creates a problem because we compute the data for `peerStore.peers`, but unmarshalling from the datastore doesn't set the peer id in the peerstore itself. So on restart we'd have no ids in the peerstore, but data in each of the books which sounds like a recipe for some edge cases in the future.

Reply:
Yeah, this does not create an issue now, but the keyBook will be populated with the PeerId :)

Reply:
Mostly what I mean is that `PeerStore.peerIds` is an unreliable Map. It might not contain all the peers we know about. `PeerStore.peers` is computed, so it will be reliable.